tokio-util/src/io/sync_read_stream.rs

```diff
@@ -0,0 +1,104 @@
+    reader: Arc<Mutex<R>>,
+    incoming_data: Receiver<Result<bytes::Bytes, io::Error>>,
+    send_data: Sender<Result<bytes::Bytes, io::Error>>,
+    rt: Handle,
+    buffer_size: usize,
+}
+
+impl<R: Read + Send> SyncReadStream<R> {
+    /// This must be called from within a Tokio runtime context, or else it will panic.
+    #[track_caller]
+    pub fn new(rt: Handle, reader: R, buffer_size: usize) -> Self {
+        let (send_data, incoming_data) = channel(buffer_size);
+
+        Self {
+            rt,
+            buffer_size,
+            reader: Arc::new(reader.into()),
+            incoming_data,
+            send_data,
+        }
+    }
+
+    /// This must be called from within a Tokio runtime context, or else it will panic.
+    pub fn new_with_handle(tokio: Handle, readable: R) -> Self {
+        Self::new(tokio, readable, DEFAULT_BUFFER_SIZE)
+    }
+
+    /// This must be called from within a Tokio runtime context, or else it will panic.
+    pub fn new_with_buffer_size(readable: R, buffer_size: usize) -> Self {
+        Self::new(Handle::current(), readable, buffer_size)
+    }
+}
+
+impl<R: Read + Send + 'static> Stream for SyncReadStream<R> {
+    type Item = Result<bytes::Bytes, io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let buffer_size = self.buffer_size;
+        // Deadlock concerns:
+        // Another `poll_next` subsequent on another thread: not a problem, one of the two will wait at
+        // this serialization point.
+        // A spawn_blocking thread reading the synchronous reader: if the read is itself "deadlocked",
+        // there's nothing we can do. As long as it returns in finite time, data will be pushed
+        match self.incoming_data.poll_recv(cx) {
+            Poll::Ready(msg) => return Poll::Ready(msg),
+            Poll::Pending => {}
+        }
+        // TODO: count when it's poisoned.
+
+        let queue_data = self.send_data.clone();
+        let reader = Arc::clone(&self.reader);
+
+        self.rt.spawn_blocking(move || {
+            let mut buf = vec![0u8; buffer_size];
+
+            let mut e_reader = reader.lock().unwrap();
+            let read_result: Result<bytes::Bytes, _> = match e_reader.read(&mut buf) {
+                Ok(n) if n == 0 => {
+                    drop(queue_data);
+                    return;
+                }
+                Ok(n) => {
+                    // This drops data from `buf` we might never see again.
+                    // There is no easy way to signal it to the caller
+                    // TODO: We should probably have a counter or something for dropped data
+                    // Including poisoning events
+                    buf.truncate(n);
+                    Ok(buf.into())
+                }
+                Err(err) => Err(err),
+            };
+
+            let _ = queue_data.blocking_send(read_result);
+        });
+
+        Poll::Pending
+    }
+}
+
```

Comment:
With this new implementation, we can also impl `FusedStream` for `SyncReadStream`:

```suggestion

impl futures_core::FusedStream for SyncReadStream {
    fn is_terminated(&self) -> bool {
        self.join_handle.is_none()
    }
}
```